Compile-time pointer sign comparison warnings in `mps_statespace.h`

### Describe the issue

Compiling the tests with `bazel build tests:all` or `bazel test tests:all` results in some warnings about signed vs unsigned integer comparisons:

```
In file included from tests/mps_statespace_test.cc:15:
tests/../lib/mps_statespace.h: In instantiation of 'static void qsim::mps::MPSStateSpace<For, FP>::ReduceDensityMatrix(MPS&, MPS&, int, fp_type*) [with For = qsim::SequentialFor; FP = float; fp_type = float]':
tests/mps_statespace_test.cc:855:25:   required from here
  855 |   ss.ReduceDensityMatrix(mps, scratch, 0, rdm);
      |   ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
tests/../lib/mps_statespace.h:244:36: warning: comparison of integer expressions of different signedness: 'int' and 'unsigned int' [-Wsign-compare]
  244 |     const bool last_index = (index == num_qubits - 1);
      |                             ~~~~~~~^~~~~~~~~~~~~~~~~~
tests/../lib/mps_statespace.h:269:28: warning: comparison of integer expressions of different signedness: 'unsigned int' and 'int' [-Wsign-compare]
  269 |     for (unsigned i = 1; i < index; ++i) {
      |                          ~~^~~~~~~
tests/../lib/mps_statespace.h:304:15: warning: comparison of integer expressions of different signedness: 'int' and 'unsigned int' [-Wsign-compare]
  304 |     if (index < num_qubits - 1) {
      |         ~~~~~~^~~~~~~~~~~~~~~~
tests/../lib/mps_statespace.h:308:41: warning: comparison of integer expressions of different signedness: 'unsigned int' and 'int' [-Wsign-compare]
  308 |     for (unsigned i = num_qubits - 2; i > index; --i) {
      |                                       ~~^~~~~~~
```


### Tell us how to reproduce the issue

Compile the tests with `bazel build tests:all` or `bazel test tests:all`

### Tell us the version of qsim or qsimcirq (if relevant)

0.23.0.dev0

### Tell us the computing environment (if relevant)

Debian Linux, G++, bazel 6.5.0

--- a/lib/mps_statespace.h
+++ b/lib/mps_statespace.h
@@ -241,7 +241,7 @@ class MPSStateSpace {
     const auto num_qubits = state.num_qubits();
     const auto bond_dim = state.bond_dim();
     const auto end = Size(state);
-    const bool last_index = (index == num_qubits - 1);
+    const bool last_index = (static_cast<unsigned>(index) == num_qubits - 1);
     const auto right_dim = (last_index ? 1 : bond_dim);
     auto offset = 0;
     fp_type* state_raw = state.get();
@@ -266,10 +266,11 @@ class MPSStateSpace {
     }
 
     // Contract all internal blocks together.
-    for (unsigned i = 1; i < index; ++i) {
+    for (unsigned i = 1; i < static_cast<unsigned>(index); ++i) {
       offset = GetBlockOffset(state, i);
 
       // reshape:
+
       new (&partial_contract2)
           MatrixMap((Complex*)(state_raw_workspace), bond_dim, 2 * bond_dim);
 
@@ -301,11 +302,11 @@ class MPSStateSpace {
 
     partial_contract.setZero();
     partial_contract(0, 0) = 1;
-    if (index < num_qubits - 1) {
+    if (static_cast<unsigned>(index) < num_qubits - 1) {
       partial_contract.noalias() = top * bot.adjoint();
     }
 
-    for (unsigned i = num_qubits - 2; i > index; --i) {
+    for (unsigned i = num_qubits - 2; i > static_cast<unsigned>(index); --i) {
       offset = GetBlockOffset(state, i);
 
       // reshape:

